backend/layers/persistence/persistence.py

```diff
@@ -497,15 +498,46 @@ def get_all_versions_for_dataset(self, dataset_id: DatasetId) -> List[DatasetVer
                 dataset_versions[i] = self._row_to_dataset_version(version, dataset, version_artifacts)
             return dataset_versions
 
-    def get_all_datasets(self) -> Iterable[DatasetVersion]:
+    def get_all_mapped_datasets_and_collections(self) -> Tuple[List[DatasetVersion], List[CollectionVersion]]:
         """
-        Returns all dataset versions in published collections.
+        Returns all mapped datasets and mapped collection versions
         """
-        active_collections = self.get_all_mapped_collection_versions()
+        active_collections = list(self.get_all_mapped_collection_versions())
         dataset_version_ids = []
         for collection in active_collections:
             dataset_version_ids.extend(collection.datasets)
-        return self._get_datasets(dataset_version_ids)
+        return list(self._get_datasets(dataset_version_ids)), active_collections
+
+    def get_all_mapped_collection_versions_with_datasets(self) -> List[CollectionVersionWithDatasets]:
+        """
+        Returns all mapped collection versions with their datasets
+        """
+        mapped_datasets, mapped_collections = self.get_all_mapped_datasets_and_collections()
+
+        datasets_by_collection_id = defaultdict(list)
+        # Construct dict of collection_id: [datasets]
+        [datasets_by_collection_id[d.collection_id.id].append(d) for d in mapped_datasets]
+
+        # Turn list of CollectionVersions into CollectionVersionsWithDatasets
+        collections_with_datasets: List[CollectionVersionWithDatasets] = []
+        for collection in mapped_collections:
+            dataset_versions = datasets_by_collection_id.get(collection.collection_id.id, [])
+            collections_with_datasets.append(
+                CollectionVersionWithDatasets(
```

Comment:
This parsing is repeated but slightly different. https://github.com/chanzuckerberg/single-cell-data-portal/blob/abd337c5c1c3e6b45005a433d29237e4e6c63969/backend/layers/persistence/persistence.py#L108-L122

We should reconcile the differences otherwise we will end up with two different shapes of `CollectionVersionWithDatasets`

Reply:
What differences are you referring to? These are slightly different situations. The code you're linking to is for transforming a row object into a CollectionVersion. I'm converting a CollectionVersion to a CollectionVersionWithDatasets.

Reply:
I see now. 

suggestion: A helper function to reduce the risk of a translation error from CollectionVersion to CollectionVersionWithDatasets would be nice. There isn't currently a reuse case at the moment, but I can see it being handy for future APIs.

Reply:
I'll introduce that as soon as we repeat this operation 👍